pkg/controller/history/controller_history.go

```diff
@@ -0,0 +1,490 @@
+	CreateControllerRevision(parent metav1.Object, revision *apps.ControllerRevision) (*apps.ControllerRevision, error)
+	// DeleteControllerRevision attempts to delete revision. If the returned error is not nil, deletion has failed.
+	DeleteControllerRevision(revision *apps.ControllerRevision) error
+	// UpdateControllerRevision updates revision such that its Revision is equal to newRevision. Implementations
+	// may retry on conflict. If the returned error is nil, the update was successful and returned ControllerRevision
+	// is valid. If the returned error is not nil, the update failed and the returned ControllerRevision is invalid.
+	UpdateControllerRevision(revision *apps.ControllerRevision, newRevision int64) (*apps.ControllerRevision, error)
+	// AdoptControllerRevision attempts to adopt revision by adding a ControllerRef indicating that the parent
+	// Object of parentKind is the owner of revision. If revision is already owned, an error is returned. If the
+	// resource patch fails, an error is returned. If no error is returned, the returned ControllerRevision is
+	// valid.
+	AdoptControllerRevision(parent metav1.Object, parentKind schema.GroupVersionKind, revision *apps.ControllerRevision) (*apps.ControllerRevision, error)
+	// ReleaseControllerRevision attempts to release parent's ownership of revision by removing parent from the
+	// OwnerReferences of revision. If an error is returned, parent remains the owner of revision. If no error is
+	// returned, the returned ControllerRevision is valid.
+	ReleaseControllerRevision(parent metav1.Object, revision *apps.ControllerRevision) (*apps.ControllerRevision, error)
+}
+
+// NewHistory returns an instance of Interface that uses client to communicate with the API Server and lister to list
+// ControllerRevisions. This method should be used to create an Interface for all scenarios other than testing.
+func NewHistory(client clientset.Interface, lister appslisters.ControllerRevisionLister) Interface {
+	return &realHistory{client, lister}
+}
+
+// NewFakeHistory returns an instance of Interface that uses informer to create, update, list, and delete
+// ControllerRevisions. This method should be used to create an Interface for testing purposes.
+func NewFakeHistory(informer appsinformers.ControllerRevisionInformer) Interface {
+	return &fakeHistory{informer.Informer().GetIndexer(), informer.Lister()}
+}
+
+type realHistory struct {
+	client clientset.Interface
+	lister appslisters.ControllerRevisionLister
+}
+
+func (rh *realHistory) ListControllerRevisions(parent metav1.Object, selector labels.Selector) ([]*apps.ControllerRevision, error) {
+	// List all revisions in the namespace that match the selector
+	history, err := rh.lister.ControllerRevisions(parent.GetNamespace()).List(selector)
+	if err != nil {
+		return nil, err
+	}
+	var owned []*apps.ControllerRevision
+	for i := range history {
+		ref := controller.GetControllerOf(history[i])
+		if ref == nil || ref.UID == parent.GetUID() {
```

Comment:
if the reference is nil no controller owns it. It is added to owned. If no controller owns it we attempt to adopt it.